File: gek-core/src/main/java/xyz/fslabo/common/data/protobuf/ProtobufBeanResolveHandler.java
package xyz.fslabo.common.data.protobuf;

import com.google.protobuf.Descriptors;
import com.google.protobuf.Message;
import xyz.fslabo.annotations.Nullable;
import xyz.fslabo.common.base.Flag;
import xyz.fslabo.common.base.JieString;
import xyz.fslabo.common.bean.BasePropertyInfo;
import xyz.fslabo.common.bean.BeanException;
import xyz.fslabo.common.bean.BeanResolver;
import xyz.fslabo.common.coll.JieColl;
import xyz.fslabo.common.invoke.Invoker;
import xyz.fslabo.common.mapper.MapperException;
import xyz.fslabo.common.reflect.JieReflect;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * {@link BeanResolver.Handler} implementation for
 * <a href="https://github.com/protocolbuffers/protobuf">Protocol Buffers</a>.
 * <p>
 * Note this handler depends on {@code protobuf libs} in the runtime.
 *
 * @author fredsuvn
 */
public class ProtobufBeanResolveHandler implements BeanResolver.Handler {

    @Override
    public @Nullable Flag resolve(BeanResolver.Context context) {
        try {
            Class<?> rawType = JieReflect.getRawType(context.getType());
            if (rawType == null) {
                return null;
            }
            //Check whether it is a protobuf object
            boolean isProtobuf = false;
            boolean isBuilder = false;
            if (Message.class.isAssignableFrom(rawType)) {
                isProtobuf = true;
            }
            if (Message.Builder.class.isAssignableFrom(rawType)) {
                isProtobuf = true;
                isBuilder = true;
            }
            if (!isProtobuf) {
                return null;
            }
            Method getDescriptorMethod = rawType.getMethod("getDescriptor");
            Descriptors.Descriptor descriptor = (Descriptors.Descriptor) getDescriptorMethod.invoke(null);
            for (Descriptors.FieldDescriptor field : descriptor.getFields()) {
                BasePropertyInfo basePropertyInfo = buildProperty(context, field, rawType, isBuilder);
                context.getProperties().put(basePropertyInfo.getName(), basePropertyInfo);
            }
            return Flag.BREAK;
        } catch (MapperException e) {
            throw e;
        } catch (Exception e) {
            throw new MapperException(e);
        }
    }

    private BasePropertyInfo buildProperty(
        BeanResolver.Context builder,
        Descriptors.FieldDescriptor field,
        Class<?> rawClass,
        boolean isBuilder
    ) throws Exception {

        String rawName = field.getName();

        //map
        if (field.isMapField()) {
            String name = rawName + "Map";
            Method getterMethod = rawClass.getMethod("get" + JieString.capitalize(name));
            List<Type> argsTypes = JieReflect.getActualTypeArguments(getterMethod.getGenericReturnType(), Map.class);
            if (JieColl.isEmpty(argsTypes)) {
                throw new BeanException("Cannot get actual argument type for " + getterMethod.getGenericReturnType() + ".");
            }
            Invoker getter = Invoker.reflectMethod(getterMethod);
            if (isBuilder) {
                Method clearMethod = rawClass.getMethod("clear" + JieString.capitalize(rawName));
                Method putAllMethod = rawClass.getMethod("putAll" + JieString.capitalize(rawName), Map.class);
                Invoker setter = new Invoker() {
                    @Override
                    public @Nullable Object invoke(@Nullable Object inst, Object... args) {
                        try {
                            clearMethod.invoke(inst);
                            return putAllMethod.invoke(inst, args);
                        } catch (InvocationTargetException e) {
                            throw new BeanException(e.getCause());
                        } catch (Exception e) {
                            throw new BeanException(e);
                        }
                    }
                };
                return new Impl(name, argsTypes.get(0), getterMethod, null, getter, setter);
            } else {
                return new Impl(name, argsTypes.get(0), getterMethod, null, getter, null);
            }
        }

        //repeated
        if (field.isRepeated()) {
            String name = rawName + "List";
            Method getterMethod = rawClass.getMethod("get" + JieString.capitalize(name));
            List<Type> argsTypes = JieReflect.getActualTypeArguments(getterMethod.getGenericReturnType(), List.class);
            if (JieColl.isEmpty(argsTypes)) {
                throw new BeanException("Cannot get actual argument type for " + getterMethod.getGenericReturnType() + ".");
            }
            Invoker getter = Invoker.reflectMethod(getterMethod);
            if (isBuilder) {
                Method clearMethod = rawClass.getMethod("clear" + JieString.capitalize(rawName));
                Method addAllMethod = rawClass.getMethod("addAll" + JieString.capitalize(rawName), Iterable.class);
                Invoker setter = new Invoker() {
                    @Override
                    public @Nullable Object invoke(@Nullable Object inst, Object... args) {
                        try {
                            clearMethod.invoke(inst);
                            return addAllMethod.invoke(inst, args);
                        } catch (InvocationTargetException e) {
                            throw new BeanException(e.getCause());
                        } catch (Exception e) {
                            throw new BeanException(e);
                        }
                    }
                };
                return new Impl(name, argsTypes.get(0), getterMethod, null, getter, setter);
            } else {
                return new Impl(name, argsTypes.get(0), getterMethod, null, getter, null);
            }
        }

        // Simple object
        Method getterMethod = rawClass.getMethod("get" + JieString.capitalize(rawName));
        Type type = getterMethod.getGenericReturnType();
        Invoker getter = Invoker.reflectMethod(getterMethod);
        if (isBuilder) {
            Method setterMethod = rawClass.getMethod("set" + JieString.capitalize(rawName), JieReflect.getRawType(type));
            Invoker setter = Invoker.reflectMethod(setterMethod);
            return new Impl(rawName, type, getterMethod, setterMethod, getter, setter);
        } else {
            return new Impl(rawName, type, getterMethod, null, getter, null);
        }
    }

    private static final class Impl implements BasePropertyInfo {

        private final String name;
        private final Type type;
        private final @Nullable Method getterMethod;
        private final @Nullable Method setterMethod;
        private final Invoker getter;
        private final @Nullable Invoker setter;

        private Impl(
            String name,
            Type type,
            @Nullable Method getterMethod,
            @Nullable Method setterMethod,
            Invoker getter,
            @Nullable Invoker setter
        ) {
            this.name = name;
            this.type = type;
            this.getterMethod = getterMethod;
            this.setterMethod = setterMethod;
            this.getter = getter;
            this.setter = setter;
        }

        @Override
        public String getName() {
            return name;
        }

        @Override
        public @Nullable Object getValue(Object bean) {
            return getter.invoke(bean);
        }

        @Override
        public void setValue(Object bean, @Nullable Object value) {
            if (setter == null) {
                throw new BeanException("Not writeable.");
            }
            setter.invoke(bean, value);
        }

        @Override
        public Type getType() {
            return type;
        }

        @Override
        public @Nullable Method getGetter() {
            return getterMethod;
        }

        @Override
        public @Nullable Method getSetter() {
            return setterMethod;
        }

        @Override
        public @Nullable Field getField() {
            return null;
        }

        @Override
        public List<Annotation> getGetterAnnotations() {
            return Collections.emptyList();
        }

        @Override
        public List<Annotation> getSetterAnnotations() {
            return Collections.emptyList();
        }

        @Override
        public List<Annotation> getFieldAnnotations() {
            return Collections.emptyList();
        }

        @Override
        public List<Annotation> getAnnotations() {
            return Collections.emptyList();
        }

        @Override
        public boolean isReadable() {
            return true;
        }

        @Override
        public boolean isWriteable() {
            return setter != null;
        }
    }
}

File: boat-plugins/boat-protobuf/src/main/kotlin/xyz/srclab/common/protobuf/ProtobufBeanResolveHandler.kt
package xyz.srclab.common.protobuf

import com.google.protobuf.Descriptors
import com.google.protobuf.Message
import com.google.protobuf.MessageLiteOrBuilder
import xyz.srclab.common.base.capitalize
import xyz.srclab.common.bean.AbstractBeanResolveHandler
import xyz.srclab.common.bean.BeanResolveContext
import xyz.srclab.common.bean.BeanResolveHandler
import xyz.srclab.common.bean.BeanTypeBuilder
import xyz.srclab.common.func.InstFunc
import xyz.srclab.common.func.InstFunc.Companion.toInstFunc
import xyz.srclab.common.func.StaticFunc.Companion.toStaticFunc
import xyz.srclab.common.reflect.getTypeSignature
import xyz.srclab.common.reflect.method
import xyz.srclab.common.reflect.methodOrNull
import xyz.srclab.common.reflect.rawClass
import java.lang.reflect.Method

/**
 * Bean resolve handler supports Protobuf types.
 *
 * @see BeanResolveHandler
 * @see AbstractBeanResolveHandler
 */
object ProtobufBeanResolveHandler : AbstractBeanResolveHandler() {

    override fun resolveAccessors(
        context: BeanResolveContext,
        builder: BeanTypeBuilder,
        getters: MutableMap<String, GetterInfo>,
        setters: MutableMap<String, SetterInfo>
    ) {
        val rawClass = builder.type.rawClass

        //Check whether it is a protobuf object
        if (!MessageLiteOrBuilder::class.java.isAssignableFrom(rawClass)) {
            return
        }

        fun addProperty(field: Descriptors.FieldDescriptor, isBuilder: Boolean) {

            fun createSetter(clearMethod: Method, addAllMethod: Method): InstFunc {
                return object : InstFunc {
                    override fun invoke(inst: Any, vararg args: Any?): Any? {
                        clearMethod.invoke(inst)
                        addAllMethod.invoke(inst, *args)
                        return null
                    }
                }
            }

            val rawName = field.name

            //map
            if (field.isMapField) {
                val name = rawName + "Map"
                val getterMethod = rawClass.methodOrNull("get${name.capitalize()}")
                if (getterMethod === null) {
                    return
                }
                val type = getterMethod.genericReturnType.getTypeSignature(Map::class.java)
                val getter = getterMethod.toInstFunc()
                getters[name] = GetterInfo(name, type, getter, null, getterMethod)

                if (isBuilder) {
                    val clearMethod = rawClass.methodOrNull("clear${rawName.capitalize()}")
                    if (clearMethod === null) {
                        throw IllegalStateException("Cannot find clear method of field: $name")
                    }
                    val putAllMethod =
                        rawClass.methodOrNull("putAll${rawName.capitalize()}", Map::class.java)
                    if (putAllMethod === null) {
                        throw IllegalStateException("Cannot find put-all method of field: $name")
                    }
                    val setter = createSetter(clearMethod, putAllMethod)
                    setters[name] = SetterInfo(name, type, setter, null, null)
                }

                return
            }

            //repeated
            if (field.isRepeated) {
                val name = rawName + "List"
                val getterMethod = rawClass.methodOrNull("get${name.capitalize()}")
                if (getterMethod === null) {
                    return
                }
                val type = getterMethod.genericReturnType.getTypeSignature(List::class.java)
                val getter = getterMethod.toInstFunc()
                getters[name] = GetterInfo(name, type, getter, null, getterMethod)

                if (isBuilder) {
                    val clearMethod = rawClass.methodOrNull("clear${rawName.capitalize()}")
                    if (clearMethod === null) {
                        throw IllegalStateException("Cannot find clear method of field: $name")
                    }
                    val addAllMethod =
                        rawClass.methodOrNull("addAll${rawName.capitalize()}", Iterable::class.java)
                    if (addAllMethod === null) {
                        throw IllegalStateException("Cannot find add-all method of field: $name")
                    }
                    val setter = createSetter(clearMethod, addAllMethod)
                    setters[name] = SetterInfo(name, type, setter, null, null)
                }

                return
            }

            // Simple object
            val getterMethod = rawClass.methodOrNull("get${rawName.capitalize()}")
            if (getterMethod === null) {
                return
            }
            val type = getterMethod.genericReturnType
            val getter = getterMethod.toInstFunc()
            getters[rawName] = GetterInfo(rawName, type, getter, null, getterMethod)

            if (isBuilder) {
                val setterMethod = rawClass.methodOrNull("set${rawName.capitalize()}", type.rawClass)
                if (setterMethod === null) {
                    throw IllegalStateException("Cannot find setter method of field: $rawName")
                }
                val setter = setterMethod.toInstFunc()
                setters[rawName] = SetterInfo(rawName, type, setter, null, setterMethod)
            }
        }

        //com.google.protobuf.Descriptors.Descriptor getDescriptor()
        val getDescriptorMethod = rawClass.getMethod("getDescriptor")
        val descriptor: Descriptors.Descriptor = getDescriptorMethod.toStaticFunc().invokeTyped()
        val isBuilder = Message.Builder::class.java.isAssignableFrom(rawClass)
        for (field in descriptor.fields) {
            addProperty(field, isBuilder)
        }

        //Add class property
        val getClassMethod = rawClass.method("getClass")
        getters["class"] = GetterInfo("class", Class::class.java, getClassMethod.toInstFunc(), null, getClassMethod)

        //break
        context.complete()
    }
}